src/command.rs

```diff
@@ -43,46 +43,82 @@ macro_rules! cmd {
 
 pub fn run_cmd(
     name: &'static str,
-    args: impl IntoIterator<Item = impl AsRef<OsStr>>,
+    args: &[&str],
     verbosity: Verbosity,
 ) -> Result<(), failure::Error> {
-    run_cmd_with_env(name, args, verbosity, std::env::vars_os())
+    run_cmd_internal(name, args, None, verbosity, true)
 }
 
 pub fn run_cmd_with_env(
     name: &'static str,
-    args: impl IntoIterator<Item = impl AsRef<OsStr>>,
+    args: &[&str],
+    envs: HashMap<OsString, OsString>,
     verbosity: Verbosity,
-    envs: impl IntoIterator<Item = (impl AsRef<OsStr>, impl AsRef<OsStr>)>,
+) -> Result<(), failure::Error> {
+    run_cmd_internal(name, args, Some(envs), verbosity, true)
+}
+
+fn run_cmd_internal(
+    name: &'static str,
+    args: &[&str],
+    envs: Option<HashMap<OsString, OsString>>,
+    verbosity: Verbosity,
+    allow_hook: bool,
 ) -> Result<(), failure::Error> {
     let (stdout, stderr) = match verbosity {
         Verbosity::Silent => (Stdio::null(), Stdio::null()),
         _ => (Stdio::inherit(), Stdio::inherit()),
     };
-    run_cmd_with_env_and_output(name, args, envs, stdout, stderr)
-}
+    let mut cmd = std::process::Command::new(if allow_hook {
+        hook_cmd(name, &args, verbosity)?
+    } else {
+        name.to_string()
+    });
+    cmd.args(args).stdout(stdout).stderr(stderr);
 
-pub fn run_cmd_with_env_and_output(
-    name: &str,
-    args: impl IntoIterator<Item = impl AsRef<OsStr>>,
-    envs: impl IntoIterator<Item = (impl AsRef<OsStr>, impl AsRef<OsStr>)>,
-    stdout: Stdio,
-    stderr: Stdio,
-) -> Result<(), failure::Error> {
-    let output = std::process::Command::new(name)
-        .args(args)
-        .envs(envs)
-        .stdout(stdout)
-        .stderr(stderr)
-        .output()
-        .map_err(|e| match e.kind() {
-            io::ErrorKind::NotFound => Error::ExecNotFound(name.to_string()).into(),
-            _ => failure::Error::from(e),
-        })?;
+    if let Some(envs) = envs {
+        cmd.envs(envs);
+    }
+    let output = cmd.output().map_err(|e| match e.kind() {
+        io::ErrorKind::NotFound => Error::ExecNotFound(name.to_string()).into(),
+        _ => failure::Error::from(e),
+    })?;
 
     if output.status.success() {
         Ok(())
     } else {
         Err(Error::ProcessExit(name.to_string(), output.status.code().unwrap()).into())
     }
 }
+
+fn hook_cmd(
+    name: &'static str,
+    args: &[&str],
+    verbosity: Verbosity,
+) -> Result<String, failure::Error> {
+    let args = args.into_iter().collect::<Vec<_>>();
+    Ok(if name == "npm" {
+        let name = std::env::var("OASIS_NPM").unwrap_or_else(|_| name.to_string());
+        let package_dir = &args[args.iter().position(|a| **a == "--prefix").unwrap() + 1];
```

Comment:
Is `--prefix` always expected to be there? Don't know the context in which this gets called and hence not sure if it will always be present.

Reply:
If you are in the package.json directory then this is not necessary I think and hence the Q.

Reply:
yes it is. I'll make a note of that